pkg/controller/controller_ref_manager.go

```diff
@@ -142,3 +143,103 @@ func (m *PodControllerRefManager) ReleasePod(pod *api.Pod) error {
 	}
 	return err
 }
+
+type ReplicaSetControllerRefManager struct {
+	rsControl          RSControlInterface
+	controllerObject   api.ObjectMeta
+	controllerSelector labels.Selector
+	controllerKind     unversioned.GroupVersionKind
+}
+
+// NewReplicaSetControllerRefManager returns a ReplicaSetControllerRefManager that exposes
+// methods to manage the controllerRef of ReplicaSets.
+func NewReplicaSetControllerRefManager(
+	rsControl RSControlInterface,
+	controllerObject api.ObjectMeta,
+	controllerSelector labels.Selector,
+	controllerKind unversioned.GroupVersionKind,
+) *ReplicaSetControllerRefManager {
+	return &ReplicaSetControllerRefManager{rsControl, controllerObject, controllerSelector, controllerKind}
+}
+
+// Classify first filters out inactive ReplicaSets, then it classify the remaining ReplicaSets
+// into three categories: 1. matchesAndControlled are the ReplicaSets whose labels
+// match the selector of the Deployment, and have a controllerRef pointing to the
+// controller 2. matchesNeedsController are the whose labels match the Deployment,
+// but don't have a controllerRef. (ReplicaSets with matching labels but with a
+// controllerRef pointing to other object are ignored) 3. controlledDoesNotMatch
+// are the ReplicaSets that have a controllerRef pointing to the controller, but their
+// labels no longer match the selector.
+func (m *ReplicaSetControllerRefManager) Classify(replicaSets []*extensions.ReplicaSet) (
+	matchesAndControlled []*extensions.ReplicaSet,
```

Comment:
I would advise against the use of named returns since it's harder to process the function if it has multiple return calls.